executor/metric_reader.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/infoschema/metricschema"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	pmodel "github.com/prometheus/common/model"
+)
+
+const promReadTimeout = time.Second * 10
+
+// MetricReaderExec uses to read metric data.
+type MetricReaderExec struct {
+	baseExecutor
+	table      *model.TableInfo
+	tblDef     *metricschema.MetricTableDef
+	outputCols []*model.ColumnInfo
+	done       bool
+}
+
+// Next implements the Executor Next interface.
+func (e *MetricReaderExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if e.done {
+		return nil
+	}
+	rows, err := e.getRows()
+	if err != nil {
+		return err
+	}
+	e.done = true
+
+	if len(rows) == 0 {
+		return nil
+	}
+
+	req.GrowAndReset(len(rows))
+	mutableRow := chunk.MutRowFromTypes(retTypes(e))
+	for _, row := range rows {
+		mutableRow.SetDatums(row...)
+		req.AppendRow(mutableRow.ToRow())
+	}
+	return nil
+}
+
+func (e *MetricReaderExec) getRows() (fullRows [][]types.Datum, err error) {
+	tblDef, err := metricschema.GetMetricTableDef(e.table.Name.L)
+	if err != nil {
+		return nil, err
+	}
+	e.tblDef = tblDef
+	// TODO: Get query range from plan instead of use default range.
+	queryRange := e.getDefaultQueryRange()
+	queryValue, err := e.queryMetric(queryRange)
+	if err != nil {
+		return nil, err
+	}
+
+	fullRows = e.genRows(queryValue, queryRange)
+	if len(e.outputCols) == len(e.table.Columns) {
+		return
+	}
+	rows := make([][]types.Datum, len(fullRows))
+	for i, fullRow := range fullRows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		rows[i] = row
+	}
+	return rows, nil
+}
+
+func (e *MetricReaderExec) queryMetric(queryRange v1.Range) (pmodel.Value, error) {
+	addr, err := e.getMetricAddr()
+	if err != nil {
+		return nil, err
+	}
+
+	queryClient, err := newQueryClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	promQLAPI := v1.NewAPI(queryClient)
+	ctx, cancel := context.WithTimeout(context.Background(), promReadTimeout)
```

Comment:
It's better to use the `ctx` received from `Next` method.

Reply:
great, done.